sdk/rust/src/contract/contract_function_parameters.rs

```diff
@@ -0,0 +1,1268 @@
+use std::cmp::max;
+use std::iter;
+use std::str::FromStr;
+
+use num_bigint::{
+    BigInt,
+    BigUint,
+    Sign,
+};
+
+use crate::contract::contract_function_selector::ContractFunctionSelector;
+use crate::evm_address::IdEvmAddress;
+
+#[derive(Debug, Clone)]
+pub struct ContractFunctionParameters {
+    args: Vec<Argument>,
+}
+
+#[derive(Debug, Clone)]
+struct Argument {
+    type_name: &'static str,
+    value_bytes: Vec<u8>,
+    is_dynamic: bool,
+}
+
+trait IntEncode {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>);
+}
+
+impl IntEncode for u8 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i8 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for u16 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i16 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for u32 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i32 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for u64 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i64 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for u128 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for i128 {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.is_negative(), self.to_be_bytes().to_vec())
+    }
+}
+
+impl IntEncode for BigUint {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (false, self.to_bytes_be())
+    }
+}
+
+impl IntEncode for BigInt {
+    fn get_is_negative_and_be_bytes(&self) -> (bool, Vec<u8>) {
+        (self.sign() == Sign::Minus, self.to_signed_bytes_be())
+    }
+}
+
+impl ContractFunctionParameters {
+    /// Create a new, empty `ContractFunctionParameters`
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        ContractFunctionParameters { args: vec![] }
+    }
+
+    fn left_pad_32_bytes(bytes: &[u8], is_negative: bool) -> Vec<u8> {
+        let pad_byte = if is_negative { 0xFF } else { 0x00 };
+        iter::repeat(pad_byte).take(32 - bytes.len()).chain(bytes.iter().map(|b| *b)).collect()
+    }
```

Comment:
```suggestion
    fn left_pad_32_bytes(bytes: &[u8], is_negative: bool) -> [u8; 32] {
        let pad_byte = if is_negative { 0xFF } else { 0x00 };
        
        let mut result = [pad_byte; 32];
        result[(32 - bytes.len())..].copy_from_slice(&bytes);
        result
    }
```
I'm not sure if it's _that_ valuable (it saves an allocation sometimes, but it's also 32 bytes, but it also happens on encoding any array) and if you want to keep the return type as `Vec` there's the `vec!` macro. This is one of those cases where the iterator chain looks weirder than the procedural version (feel free to dispute this)